module4/ex17_formatted_cart/ex17_formatted_cart.py

```diff
@@ -1,9 +1,33 @@
 class Cart:
-    pass
+    def __init__(self):
+        self.items = []
+
+    def add(self, item):
+        self.items.append(item)
+
+    def __format__(self, format_spec):
+        if format_spec == 'short':
+            new_list = sorted([item.name for item in self.items])
+            return ', '.join(new_list)
+        elif format_spec == 'long':
+            return '\n'.join(str(item) for item in sorted(self.items, key=lambda x: x.name))
+        else:
+            raise TypeError("unsupported format string passed to Cart.__format__")
 
 
 class Item:
-    pass
+    def __init__(self, quantity, measure, name, price):
+        self.quantity = quantity
+        self.measure = measure
+        self.name = name
+        if type(price) == int:
```

Comment:
There a couple of things to consider here:
1. Constructors usually raise ValueError or TypeError for constructor arguments validation.
2. Depending on assumptions made one of these errors could be raised: 

- if allowed types for price are only strictly numeric values than it's enough to check price data type to be int or float and to raise a TypeError otherwise.
- if allowed types are numeric-like values (i.e. including strings representing a number) than a conversion to float could be done within a try-except block with ValueError.

So here you could approach the solution both ways depending on the assumptions made.

Another general thing to mention is the approach to validation, there are 2 main ways:
- EAFP - "It's easier to ask for forgiveness than permission" - this coding style assumes the existence of valid keys or attributes and catches exceptions if the assumption proves false
- LBYL - "Look before you leap" - this one prefers if statements over try-except blocks.
Pythons encourages EAFP style, though sometimes LBYL suits more.